src/layer/decorators.js

```diff
@@ -0,0 +1,143 @@
+        }
+
+        var annotations = this.renderer.session.$annotations;
+        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+        if (annotations) {
+            annotations.forEach(function (item) {
+                switch (item.type) {
+                    case "info":
+                        item.typeNumber = 1;
+                        break;
+                    case "warning":
+                        item.typeNumber = 2;
+                        break;
+                    case "error":
+                        item.typeNumber = 3;
+                        break;
+                }
+            });
+            annotations = annotations.sort(compare);
+            var foldData = this.renderer.session.$foldData;
+
+            for (let i = 0; i < annotations.length; i++) {
+                let currentRow = annotations[i].row;
+                let compensateFold = this.compensateFoldRows(currentRow, foldData);
+                let currentY = Math.round((currentRow - compensateFold) * this.lineHeight * this.heightRatio);
+                let y1 = Math.round(((currentRow - compensateFold) * this.lineHeight * this.heightRatio));
+                let y2 = Math.round((((currentRow - compensateFold) * this.lineHeight + this.lineHeight) * this.heightRatio));
+                const height = y2 - y1;
+                if (height < this.minDecorationHeight) {
+                    let yCenter = ((y1 + y2) / 2) | 0;
+                    if (yCenter < this.halfMinDecorationHeight) {
+                        yCenter = this.halfMinDecorationHeight;
+                    }
+                    else if (yCenter + this.halfMinDecorationHeight > this.canvasHeight) {
+                        yCenter = this.canvasHeight - this.halfMinDecorationHeight;
+                    }
+                    y1 = Math.round(yCenter - this.halfMinDecorationHeight);
+                    y2 = Math.round(yCenter + this.halfMinDecorationHeight);
+                }
+
+                switch (annotations[i].type) {
+                    case "info":
+                        ctx.fillStyle = colors.info;
+                        break;
+                    case "warning":
+                        ctx.fillStyle = colors.warning;
+                        break;
+                    case "error":
+                        ctx.fillStyle = colors.error;
+                        break;
+                }
+                ctx.fillRect(0, currentY, this.canvasWidth, y2 - y1);
+            }
+        }
+        var cursor = this.renderer.session.selection.getCursor();
+        if (cursor) {
+            let compensateFold = this.compensateFoldRows(cursor.row, foldData);
+            let currentY = Math.round((cursor.row - compensateFold) * this.lineHeight * this.heightRatio);
+            ctx.fillStyle = "rgba(0, 0, 0, 0.5)";
+            ctx.fillRect(0, currentY, this.canvasWidth, 2);
+        }
+
+    };
+
+    this.compensateFoldRows = function (currentRow, foldData) {
```

Comment:
I'd rename currentRow to just row to avoid confusion with current cursor position